### ISSUE TITLE
SecurityContext(withTrustedRoots: true) does not use the real device trusted store and instead is using its own list from somewhere else

### ISSUE THREAD

**Issue by acoutts** [flutter/flutter#95632]
The documentation for `SecurityContext` states:
```
  /// The default security context used by most operation requiring one.
  ///
  /// Secure networking classes with an optional `context` parameter
  /// use the [defaultContext] object if the parameter is omitted.
  /// This object can also be accessed, and modified, directly.
  /// Each isolate has a different [defaultContext] object.
  /// The [defaultContext] object uses a list of well-known trusted
  /// certificate authorities as its trusted roots. On Linux and Windows, this
  /// list is taken from Mozilla, who maintains it as part of Firefox. On,
  /// MacOS, iOS, and Android, this list comes from the trusted certificates
  /// stores built in to the platforms.
  external static SecurityContext get defaultContext;
```
Specifically that on iOS / Android the list comes from the platform trust store. But what I am observing in practice is some users on corporate wifi / VPN connections are unable to use my app because although their device trusts a specific cert, it's not inside whichever trust store flutter is using, so the flutter app gets SSL connection errors.

To confirm it, you can verify that the device trust store is not used on Android. In practice, the issue affects both iOS and Android devices. 

## Steps to Reproduce

1. In the `Trusted credentials` settings, disable all system root CAs to untrust all device root certificates.
2. Open a web browser or try to use a native app and you will see SSL errors as expected
3. Open your flutter app and it will connect to the internet without any SSL errors

What you will see is any native libraries that your flutter app uses will be printing SSL errors as expected but connections established from within Dart will not respect the device trust store.

**Expected results:** 
Flutter apps should use the platform trust store.

**Actual results:** 
Flutter is using its own trust store which we cannot see.

<details>
<summary>Code sample</summary>

```dart
```

</details>

<details>
  <summary>Logs</summary>

```
```

```
```

```
[✓] Flutter (Channel stable, 2.8.0, on macOS 11.6 20G165 darwin-x64, locale en-US)
    • Flutter version 2.8.0 at /Users/andrewcoutts/fvm/versions/2.8.0
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision cf44000065 (13 days ago), 2021-12-08 14:06:50 -0800
    • Engine revision 40a99c5951
    • Dart version 2.15.0

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at /Users/andrewcoutts/Library/Android/sdk
    • Platform android-30, build-tools 30.0.2
    • ANDROID_HOME = /Users/andrewcoutts/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 13.0)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)

[✓] VS Code (version 1.52.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.12.1

[✓] VS Code (version 1.64.0-insider)
    • VS Code at /Applications/Visual Studio Code - Insiders.app/Contents
    • Flutter extension version 3.29.0

[✓] Connected device (5 available)
    • sdk gphone x86 arm (mobile) • emulator-5554                        • android-x86    • Android 11 (API 30) (emulator)
    • iPhone 13 Pro Max (mobile)  • 7823EA13-AAB6-40F2-A17D-BA16372C3D51 • ios            • com.apple.CoreSimulator.SimRuntime.iOS-15-0 (simulator)
    • macOS (desktop)             • macos                                • darwin-x64     • macOS 11.6 20G165 darwin-x64
    • Chrome (web)                • chrome                               • web-javascript • Google Chrome 96.0.4664.110
    ! Error: xff0c-13p is busy: Fetching debug symbols for xff0c-13p. Xcode will continue when xff0c-13p is finished. (code -10)

• No issues found!
```

</details>

**Comment by aam** [comment]
If you want to use locally trusted certificates you don't need to create `SecurityContext(withTrustedRoots: true)` or deal with optionally-provided `SecurityContext` instances - the default should work fine.
Also `withTrustedRoots: true` means dart vm will seed list of trusted roots with certificates that built-in into the vm and expect developers to add more certificates if needed via `setTrustedCertificates` and similar to add more trusted certificates.

**Comment by acoutts** [comment]
Yes I understand that’s how it’s documented but that doesn’t explain why it’s not working.

Can you please re-open the issue?

**Comment by aam** [comment]
> Can you please re-open the issue?

Basically `SecurityContext(withTrustedRoots: true) does not use the real device trusted store and instead is using its own list from somewhere else` is working as expected: list of certificates used when you use such `SecurityContext` are built-in into the vm.

If you have an issue where _default_ security context doesn't use device local certificates I suggest opening a new issue and including a sample that reproduces the problem.

**Comment by brianquinlan** [comment]
Hey @acoutts, 

If you did your "In the Trusted credentials settings, disable all system root CAs to untrust all device root certificates." step on Android then you actually hit a different bug - Dart does not honor those settings.

We had just concluded that this is not a significant security vulnerability before I went on holidays. I'll file a bug sometime today to fix it.

Regarding your issue, Dart loads certificates on Android from "/system/etc/security/cacerts". On iOS, it uses [`SecTrustEvaluateWithError`](https://developer.apple.com/documentation/security/2980705-sectrustevaluatewitherror?language=objc).
